public/js/util/sources-tree.js

```diff
@@ -59,59 +62,167 @@ function getURL(source) {
+    if (partA === partB) {
+      sharedParts.push(partA);
+    } else {
+      break;
+    }
+  }
+  return sharedParts;
+}
+
+function addToTree(currentNode, source) {
   const url = getURL(source);
   if (!url) {
     return;
   }
 
   const parts = url.pathname.split("/").filter(p => p !== "");
-  const isDir = (parts.length === 0 ||
-                 parts[parts.length - 1].indexOf(".") === -1);
+  const isDir = partsIsDir(parts);
   parts.unshift(url.host);
 
-  let path = "";
-  let subtree = tree;
+  let accumulatedPath = "";
+
+  for (let i = 0; i < parts.length; i++) {
+    const part = parts[i];
+    const contents = currentNode.contents;
 
-  for (let part of parts) {
     // Currently we assume that we are descending into a node with
     // children. This will fail if a path has a directory named the
     // same as another file, like `foo/bar.js/file.js`.
     //
-    // TODO: Be smarter about this, which we'll probably do when we
-    // are smarter about folders and collapsing empty ones.
-    assert(nodeHasChildren(subtree), `${subtree.name} should have children`);
-    const subpaths = subtree.contents;
+    // TODO: Be smarter about this.
+    assert(nodeHasChildren(currentNode),
+           `${currentNode.name} should have children`);
 
     // We want to sort alphabetically, so find the index where we
     // should insert this part.
-    let idx = subpaths.findIndex(subpath => {
-      return subpath.name.localeCompare(part) >= 0;
-    });
-
-    if (idx >= 0 && subpaths[idx].name === part) {
-      // A node with the same name already exists, simply traverse
-      // into it.
-      subtree = subpaths[idx];
+    let index = getSortedIndex(contents, part);
+
+    if (index >= 0 && contents[index].baseName === part) {
+      // A node with the same name already exists
+      if (contents[index].name !== contents[index].baseName) {
+        // This node is collapsed, merge in shared nodes.
+        currentNode = mergeIntoCollapsedNode(contents[index], accumulatedPath,
+                                             parts, i);
+        break;
+      } else {
+        // Continue traversing into the tree.
+        currentNode = contents[index];
+      }
     } else {
-      // No node with this name exists, so insert a new one in the
-      // place that is alphabetically sorted.
-      const node = createNode(part, path + "/" + part, []);
-      const where = idx === -1 ? subpaths.length : idx;
-      subpaths.splice(where, 0, node);
-      subtree = subpaths[where];
+      // Insert a new node, optionally collapsed, in the correct
+      // alphabetically sorted location, since a node with that name did not
+      // exist.
+      currentNode = insertNodeIntoContents(contents, accumulatedPath, parts, i,
+                                           index);
+      if (currentNode.name !== currentNode.baseName) {
+        // Jump to the end of this loop to insert a final leaf node.
+        i = parts.length - 2;
```

Comment:
This makes sense, but I wonder if we should iterate over `parts.slice(0, parts.length - 1)`, meaning don't include the last part in this loop. It's already a little weird that overwrite the contents of the final node below. This way might a little clear, and we create the leaf node explicitly and add it (now that you have `getSortedIndex`, it shouldn't be much more verbose).

I may be missing a problem with that though, and it if takes too much time don't worry about it. We can massage this code over time.